### ISSUE TITLE
Return multiple results from text query

### ISSUE THREAD

**Issue by wojtekmach** [elixir-ecto/myxql#15]
MySQL allows us to make this work:

```elixir
[
  %MySQL.Result{rows: [[1]]},
  %MySQL.Result{rows: [[2]]}
] = MyXQL.query_multi!(c.conn, "SELECT 1; SELECT 2")
```

Thus, we could do `"INSERT INTO this (...); INSERT INTO that (...)"`, so we could even replace calling `mysql` CLI when running migrations. This is only possible using the text protocol.

**Comment by wojtekmach** [comment]
@josevalim when a query returns multiple results, currently we just return the first result. This is true for text query with multiple statements and a binary query with stored procedure. (Since binary queries cannot have multiple statements, the only way they can have return multiple results is if stored procedure returns multiple)

When there are multiple results, MySQL returns a :server_more_results_exists flag and so we can easily detect this situation.

I can think of 2 ways we can handle them:

1. Document it, that if query returns multiple results, we only get the first one and so if a query may return multiple results, users should always use `query_multi!` (for text) and `stream` (for binary). 

2. Document it and return an error. We could even disconnect the connection since the server is waiting to send more results, so by disconnecting it has an opportunity to properly clean up it's state.

WDYT? I'm leaning to option 2 and we could already error telling users to use stream.

**Comment by fishcakez** [comment]
MyXQL has flags to enable these features, if going with 2 (note can close the query so disconnect not required) wouldnt it be better to unset these flags and let MySQL send back an error? This would mean not supporting multistatements/results and unsure if thats an issue or not.

**Comment by wojtekmach** [comment]
I went with option 2. But it's a good idea not to enable a capability that users cannot really use yet, I'll look into dropping them if I can. I need text multi statements to do in one query `"ROLLBACK TO SAVEPOINT myxql_savepoint; RELEASE SAVEPOINT myxql_savepoint"`, I believe I need multi results to support cursors but not sure.

**Comment by fishcakez** [comment]
Don't require to do roll back and single call. It wasn't needed for cursors
in mariaex. So shouldn't support just because of those.

**Comment by wojtekmach** [comment]
TL;DR given Mariaex has buggy implementation of handling multiple statements and results, we will explicitly not support them in MyXQL for now. I'll drop these capabilities from being announced on handshake.

cc @josevalim 

---

@fishcakez good points.

MySQL has 3 capabilities that are relevant here:
- `CLIENT_MULTI_STATEMENTS` - supports multiple statements in one query, semcolon-separated
- `CLIENT_MULTI_RESULTS` - supports multiple results for text queries (`COM_QUERY`)
- `CLIENT_PS_MULTI_RESULTS` - supports multiple results for prepared queries (`COM_STMT_EXECUTE`)

Multiple results capability is needed to handle multiple statements and calling a procedure that returns multiple results.

Btw, Mariaex sets all these 3 capability flags, although it leads to buggy behaviour:

1. Sending multiple statements in text query

   <details>

   ```
   iex(1)> {:ok, pid} = Mariaex.start_link(database: "myxql_test")
   iex(2)> Mariaex.query!(pid, "SELECT 1; SELECT 2", [], query_type: :text)
   %Mariaex.Result{
     columns: ["1"],
     connection_id: #PID<0.199.0>,
     last_insert_id: nil,
     num_rows: 1,
     rows: [[1]]
   }

   # nothing for a second or two, and then on next ping we get this crash. This is because there's another result that hasn't been consumed when executing the query.

   11:46:44.332 [error] GenServer #PID<0.199.0> terminating
   ** (FunctionClauseError) no function clause matching in Mariaex.Messages.decode_msg/2
       (mariaex) lib/mariaex/messages.ex:289: Mariaex.Messages.decode_msg(<<1>>, :running)
       (mariaex) lib/mariaex/messages.ex:253: Mariaex.Messages.decode/2
       (mariaex) lib/mariaex/protocol.ex:1240: Mariaex.Protocol.msg_decode/2
       (mariaex) lib/mariaex/protocol.ex:1208: Mariaex.Protocol.msg_recv/4
       (mariaex) lib/mariaex/protocol.ex:1248: Mariaex.Protocol.ping_recv/2
       (db_connection) lib/db_connection/connection.ex:143: DBConnection.Connection.handle_cast/2
       (connection) lib/connection.ex:810: Connection.handle_async/3
       (stdlib) gen_server.erl:637: :gen_server.try_dispatch/4
   Last message: {:"$gen_cast", {:ping, #Reference<0.757456957.125435908.159309>, %Mariaex.Protocol{binary_as: :field_type_var_string, buffer: "", cache: #Reference<0.757456957.125435908.159304>, catch_eof: false, connection_id: #PID<0.199.0>, cursors: %{}, datetime: :structs, deprecated_eof: true, json_library: Poison, lru_cache: {100, #Reference<0.757456957.125435908.159305>}, opts: [socket_options: [], sock_type: :tcp, cache_size: 100, timeout: 5000, port: 3306, hostname: "localhost", password: nil, username: "wojtek", database: "myxql_test"], seqnum: 0, sock: {Mariaex.Connection.Tcp, #Port<0.7>}, ssl_conn_state: :not_used, state: :running, state_data: nil, timeout: 5000, transaction_status: :idle}}}
   State: {Mariaex.Protocol, %Mariaex.Protocol{binary_as: :field_type_var_string, buffer: "", cache: #Reference<0.757456957.125435908.159304>, catch_eof: false, connection_id: #PID<0.199.0>, cursors: %{}, datetime: :structs, deprecated_eof: true, json_library: Poison, lru_cache: {100, #Reference<0.757456957.125435908.159305>}, opts: [socket_options: [], sock_type: :tcp, cache_size: 100, timeout: 5000, port: 3306, hostname: "localhost", password: nil, username: "wojtek", database: "myxql_test"], seqnum: 0, sock: {Mariaex.Connection.Tcp, #Port<0.7>}, ssl_conn_state: :not_used, state: :running, state_data: nil, timeout: 5000, transaction_status: :idle}}
   ```
   </details>

2. Text query that calls procedure which returns multiple results

   <details>

   ```
   iex> {:ok, pid} = Mariaex.start_link(database: "myxql_test"); Mariaex.query!(pid, "CALL multi_procedure()", [], [query_type: :text])
   %Mariaex.Result{
     columns: ["1"],
     connection_id: #PID<0.274.0>,
     last_insert_id: nil,
     num_rows: 1,
     rows: [[1]]
   }

   # nothing for a second or two, see bug above

   12:31:07.629 [error] GenServer #PID<0.274.0> terminating
   ** (FunctionClauseError) no function clause matching in Mariaex.Messages.decode_msg/2
       (mariaex) lib/mariaex/messages.ex:289: Mariaex.Messages.decode_msg(<<1>>, :running)
       (mariaex) lib/mariaex/messages.ex:253: Mariaex.Messages.decode/2
       (mariaex) lib/mariaex/protocol.ex:1240: Mariaex.Protocol.msg_decode/2
       (mariaex) lib/mariaex/protocol.ex:1208: Mariaex.Protocol.msg_recv/4
       (mariaex) lib/mariaex/protocol.ex:1248: Mariaex.Protocol.ping_recv/2
       (db_connection) lib/db_connection/connection.ex:143: DBConnection.Connection.handle_cast/2
       (connection) lib/connection.ex:810: Connection.handle_async/3
       (stdlib) gen_server.erl:637: :gen_server.try_dispatch/4
   Last message: {:"$gen_cast", {:ping, #Reference<0.1348364727.2005794818.184533>, %Mariaex.Protocol{binary_as: :field_type_var_string, buffer: <<1, 0, 0, 5, 1, 23, 0, 0, 6, 3, 100, 101, 102, 0, 0, 0, 1, 50, 0, 12, 63, 0, 1, 0, 0, 0, 8, 129, 0, 0, 0, 0, 2, 0, 0, 7, 1, 50, 7, 0, 0, 8, 254, ...>>, cache: #Reference<0.1348364727.2005794818.184528>, catch_eof: false, connection_id: #PID<0.274.0>, cursors: %{}, datetime: :structs, deprecated_eof: true, json_library: Poison, lru_cache: {100, #Reference<0.1348364727.2005794818.184529>}, opts: [socket_options: [], sock_type: :tcp, cache_size: 100, timeout: 5000, port: 3306, hostname: "localhost", password: nil, username: "wojtek", database: "myxql_test"], seqnum: 0, sock: {Mariaex.Connection.Tcp, #Port<0.18>}, ssl_conn_state: :not_used, state: :running, state_data: nil, timeout: 5000, transaction_status: :idle}}}
   State: {Mariaex.Protocol, %Mariaex.Protocol{binary_as: :field_type_var_string, buffer: "", cache: #Reference<0.1348364727.2005794818.184528>, catch_eof: false, connection_id: #PID<0.274.0>, cursors: %{}, datetime: :structs, deprecated_eof: true, json_library: Poison, lru_cache: {100, #Reference<0.1348364727.2005794818.184529>}, opts: [socket_options: [], sock_type: :tcp, cache_size: 100, timeout: 5000, port: 3306, hostname: "localhost", password: nil, username: "wojtek", database: "myxql_test"], seqnum: 0, sock: {Mariaex.Connection.Tcp, #Port<0.18>}, ssl_conn_state: :not_used, state: :running, state_data: nil, timeout: 5000, transaction_status: :idle}}
   ```

3. Binary query that calls procedure which returns multiple results

   <details>

   ```
   iex> {:ok, pid} = Mariaex.start_link(database: "myxql_test"); Mariaex.query!(pid, "CALL multi_procedure()", [], [query_type: :binary])
   ** (MatchError) no match of right hand side value: ""
       (mariaex) lib/mariaex/coder.ex:215: Mariaex.Coder.Utils.length_encoded_string/1
       (mariaex) lib/mariaex/messages.ex:211: Mariaex.Messages.decode_column_definition_41/1
       (mariaex) lib/mariaex/messages.ex:253: Mariaex.Messages.decode/2
       (mariaex) lib/mariaex/protocol.ex:1240: Mariaex.Protocol.msg_decode/2
       (mariaex) lib/mariaex/protocol.ex:1208: Mariaex.Protocol.msg_recv/4
       (mariaex) lib/mariaex/protocol.ex:667: Mariaex.Protocol.binary_query_more/5
       (db_connection) lib/db_connection/holder.ex:268: DBConnection.Holder.holder_apply/4
       (db_connection) lib/db_connection.ex:1181: DBConnection.run_execute/5
   iex(28)>
   12:32:34.853 [error] GenServer #PID<0.282.0> terminating
   ** (DBConnection.ConnectionError) client #PID<0.189.0> stopped: ** (MatchError) no match of right hand side value: ""
       (mariaex) lib/mariaex/coder.ex:215: Mariaex.Coder.Utils.length_encoded_string/1
       (mariaex) lib/mariaex/messages.ex:211: Mariaex.Messages.decode_column_definition_41/1
       (mariaex) lib/mariaex/messages.ex:253: Mariaex.Messages.decode/2
       (mariaex) lib/mariaex/protocol.ex:1240: Mariaex.Protocol.msg_decode/2
       (mariaex) lib/mariaex/protocol.ex:1208: Mariaex.Protocol.msg_recv/4
       (mariaex) lib/mariaex/protocol.ex:667: Mariaex.Protocol.binary_query_more/5
       (db_connection) lib/db_connection/holder.ex:268: DBConnection.Holder.holder_apply/4
       (db_connection) lib/db_connection.ex:1181: DBConnection.run_execute/5

       (db_connection) lib/db_connection/connection.ex:159: DBConnection.Connection.handle_cast/2
       (connection) lib/connection.ex:810: Connection.handle_async/3
       (stdlib) gen_server.erl:637: :gen_server.try_dispatch/4
       (stdlib) gen_server.erl:711: :gen_server.handle_msg/6
       (stdlib) proc_lib.erl:259: :proc_lib.wake_up/3
   Last message: {:"$gen_cast", {:stop, #Reference<0.1348364727.2005794818.186592>, %DBConnection.ConnectionError{message: "client #PID<0.189.0> stopped: ** (MatchError) no match of right hand side value: \"\"\n    (mariaex) lib/mariaex/coder.ex:215: Mariaex.Coder.Utils.length_encoded_string/1\n    (mariaex) lib/mariaex/messages.ex:211: Mariaex.Messages.decode_column_definition_41/1\n    (mariaex) lib/mariaex/messages.ex:253: Mariaex.Messages.decode/2\n    (mariaex) lib/mariaex/protocol.ex:1240: Mariaex.Protocol.msg_decode/2\n    (mariaex) lib/mariaex/protocol.ex:1208: Mariaex.Protocol.msg_recv/4\n    (mariaex) lib/mariaex/protocol.ex:667: Mariaex.Protocol.binary_query_more/5\n    (db_connection) lib/db_connection/holder.ex:268: DBConnection.Holder.holder_apply/4\n    (db_connection) lib/db_connection.ex:1181: DBConnection.run_execute/5\n"}, %Mariaex.Protocol{binary_as: :field_type_var_string, buffer: "", cache: #Reference<0.1348364727.2005794818.186587>, catch_eof: false, connection_id: #PID<0.282.0>, cursors: %{}, datetime: :structs, deprecated_eof: true, json_library: Poison, lru_cache: {100, #Reference<0.1348364727.2005794818.186588>}, opts: [socket_options: [], sock_type: :tcp, cache_size: 100, timeout: 5000, port: 3306, hostname: "localhost", password: nil, username: "wojtek", database: "myxql_test"], seqnum: 0, sock: {Mariaex.Connection.Tcp, #Port<0.20>}, ssl_conn_state: :not_used, state: :running, state_data: nil, timeout: 5000, transaction_status: :idle}}}
   State: {Mariaex.Protocol, %Mariaex.Protocol{binary_as: :field_type_var_string, buffer: "", cache: #Reference<0.1348364727.2005794818.186587>, catch_eof: false, connection_id: #PID<0.282.0>, cursors: %{}, datetime: :structs, deprecated_eof: true, json_library: Poison, lru_cache: {100, #Reference<0.1348364727.2005794818.186588>}, opts: [socket_options: [], sock_type: :tcp, cache_size: 100, timeout: 5000, port: 3306, hostname: "localhost", password: nil, username: "wojtek", database: "myxql_test"], seqnum: 0, sock: {Mariaex.Connection.Tcp, #Port<0.20>}, ssl_conn_state: :not_used, state: :running, state_data: nil, timeout: 5000, transaction_status: :idle}}
   ```
   </details>

4. Streaming results from procedure with multiple results:

   <details>

   ```
   iex(1)> {:ok, pid} = Mariaex.start_link(database: "myxql_test"); Mariaex.transaction(pid, fn conn -> Mariaex.stream(conn, "CALL multi_procedure()", []) |> Enum.to_list() end)

   ** (MatchError) no match of right hand side value: ""
       (mariaex) lib/mariaex/coder.ex:215: Mariaex.Coder.Utils.length_encoded_string/1
       (mariaex) lib/mariaex/messages.ex:211: Mariaex.Messages.decode_column_definition_41/1
       (mariaex) lib/mariaex/messages.ex:253: Mariaex.Messages.decode/2
       (mariaex) lib/mariaex/protocol.ex:1240: Mariaex.Protocol.msg_decode/2
       (mariaex) lib/mariaex/protocol.ex:1208: Mariaex.Protocol.msg_recv/4
       (mariaex) lib/mariaex/protocol.ex:667: Mariaex.Protocol.binary_query_more/5
       (mariaex) lib/mariaex/protocol.ex:913: Mariaex.Protocol.binary_first_more/5
       (mariaex) lib/mariaex/protocol.ex:835: Mariaex.Protocol.handle_fetch/4
   ```
   </details>

Given neither MyXQL nor Mariaex, as shown above, actually supports these capabilities, I'll not send them on handshake. This way, when someone attempts to use them, they'd get the following error (coming from MySQL server):

```
** (MyXQL.Error) (1312) (ER_SP_BADSELECT) PROCEDURE myxql_test.multi_procedure can't return a result set in the given context
```

Thanks again @fishcakez!

**Comment by wojtekmach** [comment]
Actually, we need `CLIENT_MULTI_RESULTS` and `CLIENT_PS_MULTI_RESULTS` to be able to use stored procedures at all (even if they send just one result), so I'll support them.

**Comment by wojtekmach** [comment]
No plans of supporting it at the moment, if someone would like to pick it up we'd of course be happy to help.
